### ISSUE TITLE
File is unknown

### ISSUE THREAD

**Issue by chrisjenx** [square/anvil#256]
Since Anvil 2.2.x I get odd file is unknown errors.

Structure is like this

// librarymodule/QueueFlushSnackbarHelper.kt
```
interface QueueFlushSnackbarHelper {
    interface Factory {
        fun create(currentRoute: Class<out Route>): QueueFlushSnackbarHelper
    }
}
```

// app module/ISPSQFlushSnackbarHelper.kt
```
class ISPSQFlushSnackbarHelper
private constructor(
    /// other imports
    private val currentRoute: Class<out Route>
) : QueueFlushSnackbarHelper {
  @ContributesBinding(AppScope::class, boundType = QueueFlushSnackbarHelper.Factory::class)
    class Factory @Inject constructor( /*injections*/ ) : QueueFlushSnackbarHelper.Factory {
        override fun create(
            currentRoute: Class<out Route>
        ): QueueFlushSnackbarHelper = ISPSQFlushSnackbarHelper(
}
```

I tried converting to assisted factory, but exact same issue:
```
@AssistedFactory
    @ContributesBinding(AppScope::class, boundType = QueueFlushSnackbarHelper.Factory::class)
    interface Factory : QueueFlushSnackbarHelper.Factory
```

Moving the factory out of the implementation class to top level package starts working, whats odd is we are doing this in other places in other library modules and it works - think those other library modules don't run dagger, so I think it's when kapt is running along side Anvil as it fails on  `kaptGenerateStubsDebugKotlin`

Stacktrace:

```
> Task :kaptGenerateStubsDebugKotlin FAILED
e: com.squareup.anvil.compiler.AnvilCompilationException: Back-end (JVM) Internal error: Couldn't resolve class for ISPSQFlushSnackbarHelper.Factory.
File is unknown

        at com.squareup.anvil.compiler.codegen.PsiUtilsKt.requireClassDescriptor(PsiUtils.kt:493)
        at com.squareup.anvil.compiler.codegen.BindingModuleGenerator$flush$$inlined$flatMap$lambda$3$1.invoke(BindingModuleGenerator.kt:210)
        at com.squareup.anvil.compiler.codegen.BindingModuleGenerator$flush$$inlined$flatMap$lambda$3$1.invoke(BindingModuleGenerator.kt:68)
        at kotlin.sequences.TransformingSequence$iterator$1.next(Sequences.kt:210)
```

Unless I'm misunderstanding, I thought Anvil ran before Kapt?

**Comment by vRallev** [comment]
From the stacktrace and sample it's unclear to me why the issue occurs. Unfortunately, without a sample there isn't much I could do.

> Unless I'm misunderstanding, I thought Anvil ran before Kapt?

It does, `kaptGenerateStubsDebugKotlin` runs before Kapt. Kapt requires these stubs as input.

**Comment by chrisjenx** [comment]
@vRallev yeah me too, it is a regression from 2.1.x -> 2.2.x as we're running 2.1.x with it working fine :confused: 

I'll see if I create a reproducable sample

**Comment by RBusarow** [comment]
This is very similar to the bug fixed [here](https://github.com/square/anvil/pull/245/files#diff-34fe21df9270241bc186a0f13c49956a79e33f445aabca49072d1c9a682fb90e).  I recognized the "File is unknown" message, which I'd seen a couple of times while troubleshooting.

Here's a really quick and dirty reproducer: https://github.com/RBusarow/AnvilReproducer/tree/file_is_unknown_reproducer

The issue is this line in `PsiElement.requireFqName()`:
https://github.com/square/anvil/blob/0548426830e0037354f75363e887b3796bbaebab/compiler/src/main/java/com/squareup/anvil/compiler/codegen/PsiUtils.kt#L230

It assumes that any `KtDotQualifiedExpression` is fully qualified and tries to resolve `ISPSQFlushSnackbarHelper.Factory` without a package name, which fails.

The simplest fix would be updating it to this:
``` Kotlin
is KtDotQualifiedExpression -> module
            .resolveClassByFqName(FqName(text), FROM_BACKEND)
            ?.let { return it.fqNameSafe }
            ?: text
```

**Comment by chrisjenx** [comment]
Oh awesome, thanks for doing that @RBusarow makes sense!

**Comment by vRallev** [comment]
@RBusarow Thanks for the sample, that helped a lot to write a test for it. Your suggested fix works great! I put up a PR #257

**Comment by RBusarow** [comment]
I'm happy to help. :)
